include/aws/common/linked_list.h

```diff
@@ -37,28 +37,122 @@ struct aws_linked_list {
 
+/**
+ * These functions need to be defined first as they are used in pre
+ * and post conditions.
+ */
+
+/**
+ * Tests if the list is empty.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_empty(const struct aws_linked_list *list) {
+    return list->head.next == &list->tail;
+}
+
+/**
+ * Checks that a linked list is valid.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_is_valid(const struct aws_linked_list *list) {
+    return list && list->head.next && list->tail.prev;
+}
+
+/**
+ * Checks that the prev of the next pointer of a node points to the node.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_node_next_is_valid(const struct aws_linked_list_node *node) {
+    return node && node->next && node->next->prev == node;
+}
+
+/**
+ * Checks that the next of the prev pointer of a node points to the node.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_node_prev_is_valid(const struct aws_linked_list_node *node) {
+    return node && node->prev && node->prev->next == node;
+}
+
+/**
+ * Checks that a linked list satisfies double linked list semantic
+ * constraints.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_is_correct(const struct aws_linked_list *list) {
+    if (!aws_linked_list_is_valid(list)) {
+        return false;
+    }
+    /* Head must reach tail by following next pointers */
+    const struct aws_linked_list_node *temp = &list->head;
+    bool head_reaches_tail = false;
+    while (temp->next) {
+        if (temp->next == &list->tail) {
+            head_reaches_tail = true;
+        }
+        temp = temp->next;
+    }
+
+    /* Tail must reach head by following prev pointers */
+    temp = &list->tail;
+    bool tail_reaches_head = false;
+    while (temp->prev) {
+        if (temp->prev == &list->head) {
+            tail_reaches_head = true;
+        }
+        temp = temp->prev;
+    }
+
+    /* The next links must be valid for all nodes (except tail) */
+    temp = &list->head;
+    bool are_next_links_valid = true;
+    while (temp != &list->tail) {
+        if (!aws_linked_list_node_next_is_valid(temp)) {
+            are_next_links_valid = false;
+        }
+        temp = temp->next;
+    }
+
+    /* The prev links must be valid for all nodes (except head) */
+    temp = &list->tail;
+    bool are_prev_links_valid = true;
+    while (temp != &list->head) {
+        if (!aws_linked_list_node_prev_is_valid(temp)) {
+            are_prev_links_valid = false;
+        }
+        temp = temp->prev;
+    }
+    return head_reaches_tail && tail_reaches_head && are_next_links_valid && are_prev_links_valid;
+    /* return true; */
+}
+
 /**
  * Initializes the list. List will be empty after this call.
  */
 AWS_STATIC_IMPL void aws_linked_list_init(struct aws_linked_list *list) {
+    AWS_PRECONDITION(list);
     list->head.next = &list->tail;
     list->head.prev = NULL;
     list->tail.prev = &list->head;
     list->tail.next = NULL;
+    AWS_POSTCONDITION(aws_linked_list_is_valid(list));
+    AWS_POSTCONDITION(aws_linked_list_is_correct(list));
```

Comment:
Is this function necessary? `_is_correct()` is never checked again, and in this case it's not really proving much.

Reply:
You are right, at the moment it doesn't prove much. Also I detected a slight problem in its implementation today that i have to fix. However, I think it will be a reasonable pre and post condition in the more complicated functions of `linked_list` (such as `push_back` and `pop_front`) and that is why I added it.

Reply:
My vote would be to merge is_valid and Is_correct then. Having both is quite confusing.

Reply:
I renamed it to `is_connected` which I think better indicates what it checks. I also merged the checks into one loop.